### ISSUE TITLE
Double indentation when class has annotated constructor in the _class header_

### ISSUE THREAD

**Issue by bric3** [pinterest/ktlint#3193]
<!-- The bug you're experiencing might have already been reported!
Please search in the [issues](https://github.com/pinterest/ktlint/issues) before creating one. -->

I have the following class in a `build.gradle.kts` (although I don't think being in Gradle script matters here). `ktlint` is double indenting the class body when there's a primary constructor.

Note, there's an `editorconfig` file that both IntelliJ and ktlint use.

The resulting format is 
1. Confusing, as we're not sure 
  * why the class is not properly "closed", 
  * if the functions and fields are declared in the correct scope, especially in Kotlin.
3. And this gets even worse as the IntelliJ IDEA formatter and ktlint are not agreeing with each other, forcing useless reformat commit for indentation.


The snippet below are shortened to highlight the issue, but code can be much more involved.


## Expected Behavior
<!---Tell us what should happen. -->
<!-- If relevant, provide a link to the relevant section in the [Kotlin Coding Conventions](https://kotlinlang.org/docs/coding-conventions.html) or [Android Kotlin Style Guide](https://developer.android.com/kotlin/style-guide).  -->

<!-- Note: Ktlint should not produce any code which conflict with IntelliJ default code formatting. So please verify that your expectation is accepted by IntelliJ default code formatting. -->

I wrote and **formatted** this class in IntelliJ IDEA, and I would expect this indentation.

```kotlin
@CacheableTask
abstract class CargoTask @Inject constructor(objects: ObjectFactory, providers: ProviderFactory)
  : DefaultTask() {
  @get:InputFiles
  @get:PathSensitive(PathSensitivity.RELATIVE)
  val sources = objects.directoryProperty().convention(cargo.mainSources)

  // etc

  @TaskAction
  fun exec() {
    // ...
  }

  private val CargoTask.cargo get() = project.extensions.getByType<CargoExtension>()
}
```

Possible, I would be OK formatting the primary constructor in a structured way.

```kotlin
@CacheableTask
abstract class CargoTask
  @Inject
  constructor(
    objects: ObjectFactory,
    providers: ProviderFactory,
  ) : DefaultTask() {
```

But clearly NOT the double-indentation of the class body.

## Observed Behavior
<!---Tell us what happens instead of the expected behavior -->
<!--- Provide the exact command which was executed but please -->
<!--- ensure that the flag "--log-level=debug" is added to the -->
<!--- command as well. Provide the output of this command. -->

ktlint double indents the class body when there's a primary constructor.

```kotlin
tasks.test {
  // some config
}

@CacheableTask
abstract class CargoTask
  @Inject
  constructor(
    objects: ObjectFactory,
    providers: ProviderFactory,
  ) : DefaultTask() {
    @get:InputFiles
    @get:PathSensitive(PathSensitivity.RELATIVE)
    val sources = objects.directoryProperty().convention(cargo.mainSources)

    // etc

    @TaskAction
    fun exec() {
      // ...
    }

    private val Task.cargo get() = project.extensions.getByType<CargoExtension>()
  }
```

## Steps to Reproduce
<!--- Provide a code example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Minimize the example to the bare -->
<!--- minimum required to reproduce the bug. Small, well written -->
<!--- example are proven to be resolved quicker. -->

Actually this can be reproduced without Gradle:

Write this class in a `MyClass.kt` file.
EDIT: I noticed there's an issue with super type #2423, but the issue I'm describing happens regardless the class extends, implements a class or not. In the following reproducer, I reproduced the wrong indentation without parent class.

```kotlin
class MyClass @Inject constructor(
  foo: String,
) {
  init {
    println(foo)
  }

  fun doSomething() {
    println("Doing something")
  }
}
```

Formatting in IntelliJ, keeps the text as is.

```shell
$ brew install ktlint
==> Auto-updating Homebrew...
Adjust how often this is run with `$HOMEBREW_AUTO_UPDATE_SECS` or disable with
`$HOMEBREW_NO_AUTO_UPDATE=1`. Hide these hints with `$HOMEBREW_NO_ENV_HINTS=1` (see `man brew`).
==> Auto-updated Homebrew!
==> Updated Homebrew from d47a45b5a2 to b3c992a0dc.
No changes to formulae or casks.

==> Fetching downloads for: ktlint
✔︎ Bottle Manifest ktlint (1.8.0)                                                                                                                                                                                                                        [Downloaded    9.8KB/  9.8KB]
✔︎ Bottle ktlint (1.8.0)                                                                                                                                                                                                                                 [Downloaded   65.3MB/ 65.3MB]
==> Pouring ktlint--1.8.0.all.bottle.tar.gz
🍺  /opt/homebrew/Cellar/ktlint/1.8.0: 5 files, 71.5MB
==> Running `brew cleanup ktlint`...
Disable this behaviour by setting `HOMEBREW_NO_INSTALL_CLEANUP=1`.
Hide these hints with `HOMEBREW_NO_ENV_HINTS=1` (see `man brew`).
$ ktlint --format                                                                                                                                                                                  took 0s 12:41:49
12:44:35.680 [main] INFO com.pinterest.ktlint.cli.internal.KtlintCommandLine -- Enable default patterns [**/*.kt, **/*.kts]
```

<details><summary>DEBUG log</summary>

```text
 ktlint --format --log-level=debug                                                                                                                                                                took 0s 12:46:03
12:46:14.962 [main] INFO com.pinterest.ktlint.cli.internal.KtlintCommandLine -- Enable default patterns [**/*.kt, **/*.kts]
12:46:14.967 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered RuleSetProviderV3 with id 'standard' in ktlint JAR
12:46:15.090 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered RuleSetProviderV3 with id 'standard' in ktlint JAR
12:46:15.096 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered ReporterProviderV2 with id 'baseline' in ktlint JAR
12:46:15.096 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered ReporterProviderV2 with id 'plain' in ktlint JAR
12:46:15.096 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered ReporterProviderV2 with id 'checkstyle' in ktlint JAR
12:46:15.096 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered ReporterProviderV2 with id 'json' in ktlint JAR
12:46:15.096 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered ReporterProviderV2 with id 'format' in ktlint JAR
12:46:15.096 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered ReporterProviderV2 with id 'html' in ktlint JAR
12:46:15.096 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered ReporterProviderV2 with id 'plain-summary' in ktlint JAR
12:46:15.097 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintServiceLoader -- Discovered ReporterProviderV2 with id 'sarif' in ktlint JAR
12:46:15.097 [main] DEBUG com.pinterest.ktlint.cli.internal.ReporterAggregator -- Initializing "plain" reporter with {plain=true, color=false, color_name=DARK_GRAY, format=true}
12:46:15.101 [main] DEBUG com.pinterest.ktlint.cli.internal.FileUtils -- Start walkFileTree from directory: 'path/to/project'
12:46:15.129 [main] DEBUG com.pinterest.ktlint.cli.internal.FileUtils -- Discovered 4 files to be processed in 27 ms
12:46:15.138 [pool-1-thread-4] DEBUG com.pinterest.ktlint.rule.engine.internal.CodeFormatter -- Starting with processing file 'settings.gradle.kts'
12:46:15.138 [pool-1-thread-3] DEBUG com.pinterest.ktlint.rule.engine.internal.CodeFormatter -- Starting with processing file 'MyClass.kt'
12:46:15.138 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.CodeFormatter -- Starting with processing file 'build.gradle.kts'
12:46:15.138 [pool-1-thread-2] DEBUG com.pinterest.ktlint.rule.engine.internal.CodeFormatter -- Starting with processing file 'build.gradle.kts'
12:46:15.361 [pool-1-thread-4] DEBUG com.pinterest.ktlint.rule.engine.internal.EditorConfigLoader -- Effective editorconfig properties for file 'path/to/project/settings.gradle.kts':
        charset: utf-8
        end_of_line: lf
        insert_final_newline: true
        indent_style: space
        indent_size: 2
        ij_continuation_indent_size: 2
        tab_width: 2
        ij_kotlin_import_nested_classes: false
        ij_kotlin_imports_layout: *, java.**, javax.**, kotlin.**, ^
        ij_kotlin_name_count_to_use_star_import: 99
        ij_kotlin_name_count_to_use_star_import_for_members: 99
        ij_kotlin_block_comment_add_space: false
        ij_kotlin_block_comment_at_first_column: false
        ij_kotlin_line_comment_add_space: true
        ij_kotlin_line_comment_add_space_on_reformat: false
        ij_kotlin_line_comment_at_first_column: false
        ij_kotlin_allow_trailing_comma: true
        ij_kotlin_allow_trailing_comma_on_call_site: true
12:46:15.361 [pool-1-thread-3] DEBUG com.pinterest.ktlint.rule.engine.internal.EditorConfigLoader -- Effective editorconfig properties for file 'path/to/project/src/main/kotlin/MyClass.kt':
        charset: utf-8
        end_of_line: lf
        insert_final_newline: true
        indent_style: space
        indent_size: 2
        ij_continuation_indent_size: 2
        tab_width: 2
        ij_kotlin_import_nested_classes: false
        ij_kotlin_imports_layout: *, java.**, javax.**, kotlin.**, ^
        ij_kotlin_name_count_to_use_star_import: 99
        ij_kotlin_name_count_to_use_star_import_for_members: 99
        ij_kotlin_block_comment_add_space: false
        ij_kotlin_block_comment_at_first_column: false
        ij_kotlin_line_comment_add_space: true
        ij_kotlin_line_comment_add_space_on_reformat: false
        ij_kotlin_line_comment_at_first_column: false
        ij_kotlin_allow_trailing_comma: true
        ij_kotlin_allow_trailing_comma_on_call_site: true
12:46:15.373 [pool-1-thread-2] DEBUG com.pinterest.ktlint.rule.engine.internal.EditorConfigLoader -- Effective editorconfig properties for file 'path/to/project/build/spotless-clean/spotlessKotlinGradle/build.gradle.kts':
        charset: utf-8
        end_of_line: lf
        insert_final_newline: true
        indent_style: space
        indent_size: 2
        ij_continuation_indent_size: 2
        tab_width: 2
        ij_kotlin_import_nested_classes: false
        ij_kotlin_imports_layout: *, java.**, javax.**, kotlin.**, ^
        ij_kotlin_name_count_to_use_star_import: 99
        ij_kotlin_name_count_to_use_star_import_for_members: 99
        ij_kotlin_block_comment_add_space: false
        ij_kotlin_block_comment_at_first_column: false
        ij_kotlin_line_comment_add_space: true
        ij_kotlin_line_comment_add_space_on_reformat: false
        ij_kotlin_line_comment_at_first_column: false
        ij_kotlin_allow_trailing_comma: true
        ij_kotlin_allow_trailing_comma_on_call_site: true
12:46:15.373 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.EditorConfigLoader -- Effective editorconfig properties for file 'path/to/project/build.gradle.kts':
        charset: utf-8
        end_of_line: lf
        insert_final_newline: true
        indent_style: space
        indent_size: 2
        ij_continuation_indent_size: 2
        tab_width: 2
        ij_kotlin_import_nested_classes: false
        ij_kotlin_imports_layout: *, java.**, javax.**, kotlin.**, ^
        ij_kotlin_name_count_to_use_star_import: 99
        ij_kotlin_name_count_to_use_star_import_for_members: 99
        ij_kotlin_block_comment_add_space: false
        ij_kotlin_block_comment_at_first_column: false
        ij_kotlin_line_comment_add_space: true
        ij_kotlin_line_comment_add_space_on_reformat: false
        ij_kotlin_line_comment_at_first_column: false
        ij_kotlin_allow_trailing_comma: true
        ij_kotlin_allow_trailing_comma_on_call_site: true
12:46:15.379 [pool-1-thread-4] DEBUG com.pinterest.ktlint.rule.engine.internal.RuleProviderSorter -- Rules will be executed in order below:
           - internal:ktlint-suppression, 
           - standard:annotation-spacing, 
           - standard:backing-property-naming, 
           - standard:binary-expression-wrapping, 
           - standard:blank-line-before-declaration, 
           - standard:blank-line-between-when-conditions, 
           - standard:chain-wrapping, 
           - standard:class-naming, 
           - standard:colon-spacing, 
           - standard:comma-spacing, 
           - standard:comment-spacing, 
           - standard:comment-wrapping, 
           - standard:condition-wrapping, 
           - standard:context-receiver-list-wrapping, 
           - standard:context-receiver-wrapping, 
           - standard:curly-spacing, 
           - standard:discouraged-comment-location, 
           - standard:dot-spacing, 
           - standard:double-colon-spacing, 
           - standard:enum-entry-name-case, 
           - standard:enum-wrapping, 
           - standard:filename, 
           - standard:final-newline, 
           - standard:fun-keyword-spacing, 
           - standard:function-expression-body, 
           - standard:function-naming, 
           - standard:function-return-type-spacing, 
           - standard:function-start-of-body-spacing, 
           - standard:function-type-modifier-spacing, 
           - standard:function-type-reference-spacing, 
           - standard:if-else-bracing, 
           - standard:if-else-wrapping, 
           - standard:import-ordering, 
           - standard:kdoc, 
           - standard:kdoc-wrapping, 
           - standard:keyword-spacing, 
           - standard:mixed-condition-operators, 
           - standard:modifier-order, 
           - standard:multiline-expression-wrapping, 
           - standard:multiline-if-else, 
           - standard:multiline-loop, 
           - standard:no-blank-line-before-rbrace, 
           - standard:no-blank-line-in-list, 
           - standard:no-blank-lines-in-chained-method-calls, 
           - standard:no-consecutive-blank-lines, 
           - standard:no-consecutive-comments, 
           - standard:no-empty-class-body, 
           - standard:no-empty-file, 
           - standard:no-empty-first-line-in-class-body, 
           - standard:no-empty-first-line-in-method-block, 
           - standard:no-line-break-after-else, 
           - standard:no-line-break-before-assignment, 
           - standard:no-multi-spaces, 
           - standard:no-trailing-spaces, 
           - standard:no-unit-return, 
           - standard:no-wildcard-imports, 
           - standard:nullable-type-spacing, 
           - standard:op-spacing, 
           - standard:package-name, 
           - standard:parameter-list-spacing, 
           - standard:parameter-list-wrapping, 
           - standard:parameter-wrapping, 
           - standard:paren-spacing, 
           - standard:property-naming, 
           - standard:property-wrapping, 
           - standard:range-spacing, 
           - standard:spacing-around-angle-brackets, 
           - standard:spacing-between-declarations-with-annotations, 
           - standard:spacing-between-declarations-with-comments, 
           - standard:spacing-between-function-name-and-opening-parenthesis, 
           - standard:square-brackets-spacing, 
           - standard:statement-wrapping, 
           - standard:string-template, 
           - standard:then-spacing, 
           - standard:try-catch-finally-spacing, 
           - standard:type-argument-comment, 
           - standard:type-argument-list-spacing, 
           - standard:type-parameter-comment, 
           - standard:type-parameter-list-spacing, 
           - standard:unary-op-spacing, 
           - standard:unnecessary-parentheses-before-trailing-lambda, 
           - standard:value-argument-comment, 
           - standard:value-parameter-comment, 
           - standard:when-entry-bracing, 
           - standard:annotation, 
           - standard:modifier-list-spacing, 
           - standard:no-single-line-block-comment, 
           - standard:wrapping, 
           - standard:no-semi, 
           - standard:class-signature, 
           - standard:function-signature, 
           - standard:argument-list-wrapping, 
           - standard:chain-method-continuation, 
           - standard:function-literal, 
           - standard:trailing-comma-on-call-site, 
           - standard:trailing-comma-on-declaration-site, 
           - standard:indent, 
           - standard:block-comment-initial-star-alignment, 
           - standard:string-template-indent, 
           - standard:max-line-length
12:46:15.435 [pool-1-thread-4] DEBUG com.pinterest.ktlint.rule.engine.internal.CodeFormatter -- Finished with processing file 'settings.gradle.kts'
12:46:15.443 [pool-1-thread-3] DEBUG com.pinterest.ktlint.rule.engine.internal.CodeFormatter -- Finished with processing file 'MyClass.kt'
12:46:15.529 [pool-1-thread-1] DEBUG com.pinterest.ktlint.rule.engine.internal.CodeFormatter -- Finished with processing file 'build.gradle.kts'
12:46:15.529 [pool-1-thread-2] DEBUG com.pinterest.ktlint.rule.engine.internal.CodeFormatter -- Finished with processing file 'build.gradle.kts'
12:46:15.530 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintCommandLine -- Finished processing in 567ms / 4 file(s) scanned / 11 error(s) found
12:46:15.530 [main] DEBUG com.pinterest.ktlint.cli.internal.KtlintCommandLine -- Exit ktlint with exit code: 0
```

</details>


## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Version of ktlint used: 1.8.0
* Relevant parts of the `.editorconfig` settings

   ```editorconfig
   [{*.kt,*.kts}]
   ij_kotlin_import_nested_classes = false
   ij_kotlin_imports_layout = *, java.**, javax.**, kotlin.**, ^

   ij_kotlin_name_count_to_use_star_import = 99
   ij_kotlin_name_count_to_use_star_import_for_members = 99

   ij_kotlin_block_comment_add_space = false
   ij_kotlin_block_comment_at_first_column = false
   ij_kotlin_line_comment_add_space = true
   ij_kotlin_line_comment_add_space_on_reformat = false
   ij_kotlin_line_comment_at_first_column = false
   ij_kotlin_allow_trailing_comma = true
   ij_kotlin_allow_trailing_comma_on_call_site = true
   ```

* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task): Spotless 8.0.0
* Version of Gradle used (if applicable): 9.2.1
* Operating System and version: macOs 15.7.2

**Comment by bric3** [comment]
Additionally:

* This makes code inconsistent has some classes are properly formatted, while some others are not.
* Not to mention the CI issues, that fails due to ktlint doing his own thing.

**Comment by bric3** [comment]
So the workaround, is to avoid _ktlint default style_, and rely on the intellij style.

Either via the `.editorconfig` file

https://pinterest.github.io/ktlint/latest/rules/code-styles/

```editorconfig
[*.{kt,kts}]
ktlint_code_style = intellij_idea
```

Or if you are using `ktlint` via spotless, by the override.

```diff
 spotless {
   kotlin {
     ktlint("1.8.0")
+      .editorConfigOverride(mapOf("ktlint_code_style" to "intellij_idea"))
   }

   kotlinGradle {
     target("**/*.gradle.kts")
     ktlint("1.8.0")
+      .editorConfigOverride(mapOf("ktlint_code_style" to "intellij_idea"))
   }
 }
```

**Comment by paul-dingemans** [comment]
Thank you for your elaborate effort to document the issue.

According to your [comment](https://github.com/pinterest/ktlint/issues/2423#issuecomment-3570437664) you have found a way around this. There are more discussions to be found in the issue tracker related. The most noteworthy, but please check other references in that issue as well, is https://github.com/pinterest/ktlint/issues/2138.

Bottom line of all discussion is that this is not going to be changed in ktlint official.

**Comment by bric3** [comment]
That's sad, but I understand.
